src/bsp/src/stm32/src/SpiEsp.cpp

```diff
@@ -0,0 +1,211 @@
+        }
+        m_inboundMessage.m_sizeBytes = 0;
+        m_rxState = receiveState::RECEIVING_HEADER;
+        rxLengthBytes = EspHeader::sizeBytes;
+        break;
+    case receiveState::ERROR:
+        rxLengthBytes = EspHeader::sizeBytes;
+        m_rxState = receiveState::RECEIVING_HEADER;
+        break;
+    }
+    if (m_inboundRequest && m_txState == transmitState::IDLE) {
+        m_txState = transmitState::SENDING_HEADER;
+    }
+
+    // Transmitting state machine
+    switch (m_txState) {
+    case transmitState::IDLE:
+        break;
+    case transmitState::SENDING_HEADER:
+        updateOutboundHeader();
+        txLengthBytes = EspHeader::sizeBytes;
+        txBuffer = (uint8_t*)&m_outboundHeader;
+        break;
+    case transmitState::SENDING_PAYLOAD:
+        txLengthBytes = m_outboundMessage.m_sizeBytes;
+        txBuffer = m_outboundMessage.m_data.data();
+        m_outboundMessage.m_sent = true;
+        break;
+    case transmitState::ERROR:
+        HAL_GPIO_WritePin(ESP_CS_GPIO_Port, ESP_CS_Pin, GPIO_PIN_SET);
+        break;
+    }
+
+    if ((m_inboundRequest || m_outboundMessage.m_sizeBytes != 0) &&
+        m_txState != transmitState::ERROR && m_rxState != receiveState::ERROR) {
+        HAL_GPIO_WritePin(ESP_CS_GPIO_Port, ESP_CS_Pin, GPIO_PIN_RESET);
+        vTaskDelay(20);
+        uint32_t finalSize = std::max(txLengthBytes, rxLengthBytes);
+        memset(m_inboundMessage.m_data.data(), 0, ESP_SPI_MAX_MESSAGE_LENGTH);
+        if (xSemaphoreTake(m_semaphore, (TickType_t)10) == pdTRUE) {
+            EspSpi_TransmitReceiveDma(txBuffer, m_inboundMessage.m_data.data(), finalSize,
+                                      SpiEsp::espTxRxCallback, this);
+            xSemaphoreGive(m_semaphore);
+        }
+    }
+}
+
+void SpiEsp::updateOutboundHeader() {
+    // TODO: get actual system state
+    m_outboundHeader.rxSizeWord = m_inboundMessage.m_sizeBytes >> 2;
+    m_outboundHeader.txSizeWord = m_outboundMessage.m_sizeBytes >> 2;
+    m_outboundHeader.crc8 = m_crc.calculateCRC8(&m_outboundHeader, EspHeader::sizeBytes - 1);
+    if (m_outboundHeader.txSizeWord == 0) {
+        m_isBusy = false;
+    }
+}
+
+void SpiEsp::espInterruptCallback(void* instance) {
+    auto* _this = static_cast<SpiEsp*>(instance);
+    // Interrupt is on both falling edge and rising edge.
+    _this->m_inboundRequest = HAL_GPIO_ReadPin(ESP_CS_GPIO_Port, ESP_CS_Pin);
+}
+
+void SpiEsp::espTxRxCallback(void* instance) {
+    auto* _this = static_cast<SpiEsp*>(instance);
+
+    HAL_GPIO_WritePin(ESP_CS_GPIO_Port, ESP_CS_Pin, GPIO_PIN_SET);
+    switch (_this->m_rxState) {
+    case receiveState::RECEIVING_HEADER:
+        _this->m_rxState = receiveState::PARSING_HEADER;
+        break;
+    case receiveState::RECEIVING_PAYLOAD:
+        _this->m_rxState = receiveState::VALIDATE_CRC;
+        break;
+    default:
+        // other states should not be present on callback
```

Comment:
In fact, the state WILL never be other than the two provided i the switch case. No could would be called in the default segment. Also, this function is called from the ISR, so logging something is probably not recommended

Reply:
Exactly, you should log a warning/error since this point is not supposed to be reached. And That's true that a function called in an ISR is not a good idea, though it shoudn't be called here anyway. It's just so it's easy to find a bug if there's one that pops up.